packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -1029,6 +1064,26 @@ class RenderEditable extends RenderBox {
     return null;
   }
 
+  /// Clamps a local coordinate within the viewport.
+  ///
+  /// This is used for example to prevent the text selection handles from
+  /// displaying outside the viewport while fading out.
+  Offset clampToViewport(Offset point) {
+    switch (_viewportAxis) {
+      case Axis.horizontal:
+        final double dx = point.dx.clamp(0.0, _viewportExtent);
+        if (dx != point.dx)
+          return Offset(point.dx.clamp(0.0, _viewportExtent), point.dy);
```

Comment:
As far as I understand, this isn't possible currently if the input coordinate is coming from `RenderEditable.getEndpointsForSelection`, because the text only scrolls in one direction (either vertically or horizontally, but not both), and therefore a handle can only be outside the viewport along one axis.

However, I realized that the code is simpler is I just use `size.contains` from the enclosing `RenderEditable`, so I've just removed this function altogether.